src/Shared/TagTransformerJsonHelper.cs

```diff
@@ -0,0 +1,190 @@
+// <copyright file="TagTransformerJsonHelper.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Buffers;
+using System.Text;
+using System.Text.Json;
+
+namespace OpenTelemetry.Internal;
+
+#nullable enable
+
+internal static class TagTransformerJsonHelper
+{
+    internal static string JsonSerializeArrayTag(Array array)
+        => array switch
+        {
+            // This switch needs to support the same set of types as those in TagTransformer.TransformArrayTagInternal
+            char[] charArray => WriteToString(charArray, WriteCharValue),
+            string[] stringArray => WriteToString(stringArray, WriteStringValue),
+            bool[] booleanArray => WriteToString(booleanArray, WriteBooleanValue),
+
+            // Runtime allows casting byte[] to sbyte[] and vice versa, so pattern match to sbyte[] doesn't work since it goes through byte[]
+            // Similarly (array is sbyte[]) doesn't help either. Only real type comparison works.
+            // This is true for byte/sbyte, short/ushort, int/uint and so on
+            byte[] byteArray => array.GetType() == typeof(sbyte[]) ? WriteToString((sbyte[])array, WriteSByteValue) : WriteToString(byteArray, WriteByteValue),
+            short[] shortArray => array.GetType() == typeof(ushort[]) ? WriteToString((ushort[])array, WriteUShortValue) : WriteToString(shortArray, WriteShortValue),
+            int[] intArray => array.GetType() == typeof(uint[]) ? WriteToString((uint[])array, WriteUIntValue) : WriteToString(intArray, WriteIntValue),
+
+            long[] longArray => WriteToString(longArray, WriteLongValue),
```

Comment:
does the above casting issue exist for `ulong`?

Reply:
Yes - it does - the original code didn't handle ulong specifically but due to the default casting behavior it probably worked regardless. I added ulong handling as well.